### ISSUE TITLE
Fatal `ImportError` in `0.7.20`

### ISSUE THREAD

**Issue by jorenham** [hukkin/mdformat#501]
### Describe the bug

**context**
running `mdformat <anything>` crashes on invocation

**expectation**
No crashes

**bug**

```console
$ mdformat README.md
Traceback (most recent call last):
  File "/home/joren/Workspace/scipy-stubs/.venv/bin/mdformat", line 5, in <module>
    from mdformat.__main__ import run
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/__init__.py", line 4, in <module>
    from mdformat._api import file, text
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/_api.py", line 10, in <module>
    from mdformat._util import EMPTY_MAP, NULL_CTX, build_mdit, detect_newline_type
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/_util.py", line 14, in <module>
    import mdformat.plugins
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/plugins.py", line 91, in <module>
    PARSER_EXTENSIONS, _PARSER_EXTENSION_DISTS = _load_entrypoints(
                                                 ~~~~~~~~~~~~~~~~~^
        importlib_metadata.entry_points(group="mdformat.parser_extension")
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/plugins.py", line 25, in _load_entrypoints
    loaded_ifaces[ep.name] = ep.load()
                             ~~~~~~~^^
  File "/home/joren/.pyenv/versions/3.13.1/lib/python3.13/importlib/metadata/__init__.py", line 179, in load
    module = import_module(match.group('module'))
  File "/home/joren/.pyenv/versions/3.13.1/lib/python3.13/importlib/__init__.py", line 88, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat_tables/__init__.py", line 5, in <module>
    from .plugin import (  # noqa: F401
    ...<4 lines>...
    )
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat_tables/plugin.py", line 5, in <module>
    from mdformat.renderer import RenderContext, RenderTreeNode
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/renderer/__init__.py", line 20, in <module>
    from mdformat.renderer._context import DEFAULT_RENDERERS, WRAP_POINT, RenderContext
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/renderer/_context.py", line 15, in <module>
    from mdformat._util import cached_textwrapper
ImportError: cannot import name 'cached_textwrapper' from partially initialized module 'mdformat._util' (most likely due to a circular import) (/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/_util.py)
```

### Reproduce the bug

`$ mdformat`

### List your environment

```
❯ mdformat --version
Traceback (most recent call last):
  File "/home/joren/Workspace/scipy-stubs/.venv/bin/mdformat", line 5, in <module>
    from mdformat.__main__ import run
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/__init__.py", line 4, in <module>
    from mdformat._api import file, text
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/_api.py", line 10, in <module>
    from mdformat._util import EMPTY_MAP, NULL_CTX, build_mdit, detect_newline_type
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/_util.py", line 14, in <module>
    import mdformat.plugins
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/plugins.py", line 91, in <module>
    PARSER_EXTENSIONS, _PARSER_EXTENSION_DISTS = _load_entrypoints(
                                                 ~~~~~~~~~~~~~~~~~^
        importlib_metadata.entry_points(group="mdformat.parser_extension")
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/plugins.py", line 25, in _load_entrypoints
    loaded_ifaces[ep.name] = ep.load()
                             ~~~~~~~^^
  File "/home/joren/.pyenv/versions/3.13.1/lib/python3.13/importlib/metadata/__init__.py", line 179, in load
    module = import_module(match.group('module'))
  File "/home/joren/.pyenv/versions/3.13.1/lib/python3.13/importlib/__init__.py", line 88, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat_tables/__init__.py", line 5, in <module>
    from .plugin import (  # noqa: F401
    ...<4 lines>...
    )
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat_tables/plugin.py", line 5, in <module>
    from mdformat.renderer import RenderContext, RenderTreeNode
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/renderer/__init__.py", line 20, in <module>
    from mdformat.renderer._context import DEFAULT_RENDERERS, WRAP_POINT, RenderContext
  File "/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/renderer/_context.py", line 15, in <module>
    from mdformat._util import cached_textwrapper
ImportError: cannot import name 'cached_textwrapper' from partially initialized module 'mdformat._util' (most likely due to a circular import) (/home/joren/Workspace/scipy-stubs/.venv/lib/python3.13/site-packages/mdformat/_util.py)
```

yea...

it's `0.7.20`

**Comment by jorenham** [comment]
I'd advice yanking 0.7.20

**Comment by hukkin** [comment]
Thanks for the issue!

Do you have plugins installed?

I can't reproduce:

```console
foo@bar:~$ pipx run 'mdformat==0.7.20' --version
mdformat 0.7.20
```

```console
foo@bar:~$ printf -- 'blaa\n====\n\n\nHello hello' > f.md
foo@bar:~$ pipx run 'mdformat==0.7.20' f.md 
foo@bar:~$ cat f.md 
# blaa

Hello hello
foo@bar:~$
```

**Comment by jorenham** [comment]
`mdformat-gfm==0.4.1` and `mdformat-gfm-alerts==1.0.1`

**Comment by hukkin** [comment]
I think any parser extension plugin is enough to cause this. I've yanked the release, thanks for reporting!

**Comment by jorenham** [comment]
> One thing I noticed is that many plugins import types that are only used for type checking. I would update documentation to encourage users using these types in TYPE_CHECKING blocks. This also improve module load time.
> 
> ```
> if TYPE_CHECKING:
>     from mdformat.renderer.typing import Render
> 
> ...
> 
> RENDERERS: MutableMapping[str, "Render"] = {}
> ```

It's not the prettiest, but you could also use `from __future__ import annotations` which allows you to use a non-stringified `MutableMapping[str, Render]` (the import will automatically stringify them instead)

**Comment by nfelt14** [comment]
> I think any parser extension plugin is enough to cause this. I've yanked the release, thanks for reporting!

Could you remove the tag in git as well? `pre-commit` autoupdated and I ran into this issue.

**Comment by hukkin** [comment]
This is fixed in `mdformat==0.7.21`
